homework/shared_ptr/test.cpp

```diff
@@ -0,0 +1,222 @@
+#include <string>
+#include "gtest/gtest.h"
+#include "control_block.hpp"
+#include "shared_ptr.hpp"
+#include "weak_ptr.hpp"
+
+const std::string testString{"Ala ma kota"};
+constexpr int testValueOne = 10;
+constexpr int testValueTwo = 20;
+const std::string deleterString = "Deleted";
+constexpr int testNumber = 0;
+
+struct A {
+    int a;
+    std::string b;
+};
+
+std::function<void(A* ptr)> deleter = [](A* ptr) { ptr->a = 0; ptr->b = "Deleted"; };
+
+struct sharedPtrTest : ::testing::Test {
+    sharedPtrTest()
+    : sPtr(new int{testValueOne}) {}
+    cs::shared_ptr<int> sPtr;
+};
+
+TEST_F(sharedPtrTest, testCounter) {
+    ASSERT_EQ(sPtr.use_count(), 1);
+    {
+        cs::shared_ptr<int> testPtr(sPtr);
+        ASSERT_EQ(sPtr.use_count(), 2);
+    };
+    ASSERT_EQ(sPtr.use_count(), 1);
+}
+
+TEST_F(sharedPtrTest, testCopyConstructor) {
+    ASSERT_EQ(*sPtr, testValueOne);
+    ASSERT_EQ(sPtr.use_count(), 1);
+    auto sPtr2(sPtr);
+    auto sPtr3 = sPtr;
+    ASSERT_EQ(*sPtr, *sPtr2);
+    ASSERT_EQ(*sPtr, *sPtr3);
+    ASSERT_EQ(sPtr3.use_count(), 3);
+}
+
+TEST_F(sharedPtrTest, testCopyAssignment) {
+    ASSERT_EQ(*sPtr, testValueOne);
+    ASSERT_EQ(sPtr.use_count(), 1);
+    cs::shared_ptr<int> sPtr2;
+    sPtr2 = sPtr;
+    cs::shared_ptr<int> sPtr3;
+    sPtr3 = sPtr;
+    ASSERT_EQ(*sPtr, *sPtr2);
+    ASSERT_EQ(*sPtr, *sPtr3);
+    ASSERT_EQ(sPtr3.use_count(), 3);
+}
+
+TEST_F(sharedPtrTest, testMoveConstructor) {
+    auto sPtr2(std::move(sPtr));
+    ASSERT_EQ(*sPtr2, testValueOne);
+    auto sPtr3 = std::move(sPtr2);
+    ASSERT_EQ(*sPtr3, testValueOne);
+    ASSERT_EQ(sPtr3.use_count(), 1);
+}
+
+TEST_F(sharedPtrTest, testMoveAssignment) {
+    cs::shared_ptr<int> sPtr2; 
+    sPtr2 = std::move(sPtr);
+    ASSERT_EQ(*sPtr2, testValueOne);
+    ASSERT_EQ(sPtr2.use_count(), 1);
+}
+
+TEST_F(sharedPtrTest, testGet) {
+    auto ptr = sPtr.get();
+    auto ptr2 = sPtr.get();
+    ASSERT_EQ(*ptr, *ptr2);
+}
+
+TEST_F(sharedPtrTest, testReset) {
+    sPtr.reset(new int{testValueTwo});
+    ASSERT_EQ(*sPtr, testValueTwo);
+    sPtr.reset();
+    ASSERT_EQ(sPtr.get(), nullptr);
+}
+
+TEST_F(sharedPtrTest, testSwap) {
+    cs::shared_ptr<int> testPtr(new int(testValueTwo));
+    cs::shared_ptr<int> counterIncreaser(sPtr);
+    auto higherUseCount = sPtr.use_count();
+    auto lowerUseCount = testPtr.use_count();
+    testPtr.swap(sPtr);
+    ASSERT_EQ(*sPtr, testValueTwo);
+    ASSERT_EQ(*testPtr, testValueOne);
+    ASSERT_EQ(testPtr.use_count(), higherUseCount);
+    ASSERT_EQ(sPtr.use_count(), lowerUseCount);
+}
+
+TEST_F(sharedPtrTest, testAccessOperator) {
+    cs::shared_ptr<std::string> uPtr2(new std::string{testString});
+    ASSERT_EQ(uPtr2->at(0), 'A');
+    ASSERT_EQ(uPtr2->at(1), 'l');
```

Comment:
```suggestion
    ASSERT_EQ(uPtr2->at(0), testString[0]);
    ASSERT_EQ(uPtr2->at(1), testString[1]);
```
Write code in such way that will require you less changes in future. With my bersion it will work even if you change the content of testString.

Reply:
done